src/lager.erl

```diff
@@ -150,28 +207,39 @@ trace_file(File, Filter, Options) when is_list(Options) ->
     trace_file(File, Filter, debug, Options).
 
 trace_file(File, Filter, Level, Options) ->
-    Trace0 = {Filter, Level, {lager_file_backend, File}},
-    case lager_util:validate_trace(Trace0) of
-        {ok, Trace} ->
-            Handlers = gen_event:which_handlers(lager_event),
+    FileName = lager_util:expand_path(File),
+    case validate_trace_filters(Filter, Level, {lager_file_backend, FileName}) of
+        {Sink, {ok, Trace}} ->
+            Handlers = lager_config:global_get(handlers, []),
             %% check if this file backend is already installed
-            Res = case lists:member({lager_file_backend, File}, Handlers) of
-               false ->
-                     %% install the handler
-                    LogFileConfig = lists:keystore(level, 1, lists:keystore(file, 1, Options, {file, File}), {level, none}),
-                    supervisor:start_child(lager_handler_watcher_sup,
-                        [lager_event, {lager_file_backend, File}, LogFileConfig]);
-                _ ->
-                    {ok, exists}
+            Res = case lists:keyfind({lager_file_backend, FileName}, 1, Handlers) of
+                      false ->
+                          %% install the handler
+                          LogFileConfig =
+                              lists:keystore(level, 1,
+                                             lists:keystore(file, 1,
+                                                            Options,
+                                                            {file, FileName}),
+                                             {level, none}),
+                          HandlerInfo =
+                              lager_app:start_handler(Sink, {lager_file_backend, FileName},
+                                                      LogFileConfig),
+                          lager_config:global_set(handlers, [HandlerInfo|Handlers]),
+                          {ok, installed};
+                      {_Watcher, _Handler, Sink} ->
+                          {ok, exists};
+                      {_Watcher, _Handler, _OtherSink} ->
+                          {error, file_in_use}
             end,
             case Res of
               {ok, _} ->
-                add_trace_to_loglevel_config(Trace),
-                {ok, {{lager_file_backend, File}, Filter, Level}};
+                %% XXX Double-check this logic for {ok, exists}
```

Comment:
Not sure exactly what you're asking here, but have you double-checked this logic for {ok, exists}? If so, kill the comment?